task_03_oop.py

```diff
@@ -0,0 +1,388 @@
+)
+print("\nCompany 101 employees:")
+for emp in company1.employees:
+    print(emp)
+
+print("\nCompany 202 employees:")
+for emp in company2.employees:
+    print(emp)
+
+all_employees = company1.employees + company2.employees
+all_ids = [emp.emp_id for emp in all_employees]
+if len(all_ids) == len(set(all_ids)):
+    print("There are no duplicated ids'")
+else:
+    print("There are duplicated ids'")
+
+# Attempting to hire the same employee twice.
+print("Attempting to hire the same employee twice".center(150, "-"))
+print(company2.hire(employee4))
+print(company2.hire(employee4))
+
+# Remove employees from the company using the fire method. Verify that employees are no longer in the company's list of employees.
+print("Remove employees from the company using the fire method".center(150, "-"))
+print(company1.fire(employee1))
+print(company2.fire(employee4))
+if (employee1 in company1.employees) or (employee4 in company2.employees):
+    print("There is an issue with firing employees")
+else:
+    print("Employees are no longer in the company's list of employees")
+
+# Attempting to fire an employee not employed by the company.
+# Verify that attempting to fire an employee not employed by the company doesn't affect the employee list.
+print("Attempting to fire an employee not employed by the company".center(150, "-"))
+print(company1.fire(employee1))
+print("\nCompany 101 employees:")
+for emp in company1.employees:
+    print(emp)
+
+print("\nCompany 202 employees:")
+for emp in company2.employees:
+    print(emp)
+
+# Set and get salaries/hourly_rates for various employees;
+print("Set and get salaries/hourly_rates for various employees".center(150, "-"))
+print(f"Employee hourly rate is {employee1.hourly_rate}")
+employee1.hourly_rate = 80
+print(f"Employee hourly rate was changed to {employee1.hourly_rate}")
+print(f"\nEmployee salary is {employee2.salary}")
+employee2.salary = 555
+print(f"Employee salary was changed to {employee2.salary}")
+
+# Call the calculate_payment method and verify that the payment is calculated correctly.
+print(
+    "Call the calculate_payment method and verify that the payment is calculated correctly".center(
+        150, "-"
+    )
+)
+print(
+    f"{employee1.name}'s hourly rate is {employee1.hourly_rate}, so if we calculate payment it will be {employee1.calculate_payment()}"
+)
+print(
+    f"{employee4.name}'s salary is {employee4.salary}, so if we calculate payment it will be {employee4.calculate_payment()}"
+)
+
+# Increase the salary or hourly_rate of an employee using the raise_pay method.
+print(
+    "Increase the salary or hourly_rate of an employee using the raise_pay method".center(
+        150, "-"
+    )
+)
+print(f"Employee's salary before raising: {employee2.salary}")
+print(company1.raise_pay(employee2, 15))
+print(f"Employee's salary after raising: {employee2.salary}")
+
+print(f"\nEmployee's hourly rate before raising: {employee3.hourly_rate}")
+print(company1.raise_pay(employee3, 15))
```

Comment:
`employee3` is hired by `company1`, but if I change the code to:

```python
print(f"\nEmployee's hourly rate before raising: {employee3.hourly_rate}")
print(company2.raise_pay(employee3, 15))
print(f"Employee's hourly rate after raising: {employee3.hourly_rate}")
```

the salary is still increased. This is an error. A company must only be able to raise pay for employees that are actually employed by it.